### ISSUE TITLE
Decouple ProcessingResult from Log and Command writer

### ISSUE THREAD

**Issue by Zelldon** [camunda/zeebe#9998]
**Description**

Currently, the interface of the `ProcessingResult` looks like this

```
public interface ProcessingResult {

  long writeRecordsToStream(LogStreamBatchWriter logStreamBatchWriter);

  boolean writeResponse(CommandResponseWriter commandResponseWriter);

  /**
   * @return <code>false</code> to indicate that the side effect could not be applied successfully
   */
  boolean executePostCommitTasks();
}
```

I would like to completely hide the Log and Command writers from the Engine (to further decouple the engine from the logstream), which means we should remove that from the interface. I would like to have a method which gives use the resulting records or a buffer, which we can internally (in the StreamProcessor) write to our LogStreamWriter.

_I see the implementation of the ProcessingResult as part of the Engine, and the interface part of the StreamProcessor._

**Comment by pihme** [comment]
> I see the implementation of the ProcessingResult as part of the Engine, and the interface part of the StreamProcessor.

I would like to discuss this next week, if possible.

Fully agree with overhauling this interface. However, I don't see this as part of the Engine. In my mind, the engine implements `RecordProcessor`. Record processor is called by state machine, which also provides the `ProcessingResultBuilder`. So then the engine is expected to use the builder to produce the result. Whoever calls the method, must provide the result builder and by doing so also has full control over the result being built.

I think one can drive the point home by changing `RecordProcessor` to
```java
<RESULT> RESULT process(TypedRecord record, ProcessingResultBuilder<RESULT> processingResultBuilder);
```

And maybe adding a method `createSkipRecordResult()` to `ProcessingResultBuilder` to avoid `EmptyProcessingResult`

Then the engine doesn't even have an interface it could implement.

Letting the engine implement the interface is dangerous in my mind. Returning a list of records, would be fine because this is an abstraction that the engine knows about. But what about the response to the client?
Returning something like a byte buffer is a big problem in my mind, cause then the engine needs to know how to create the byte buffer the right way. This introduces a heavy specification and format dependency, which is totally obscured by the generic byte buffer interface.

Also, I am thinking of having different `ProcessingResultBuilder` implementations. One for production where we can performance-optimize the heck out of it, so it doesn't copy bytes more often than necessary. And another for tests. And then the one for tests would copy all records, and give me a nice list of records that I can assert against.

So, bottom line, my proposal would be to get rid of this interface here and make it an implementation detail of the implementation of `ProcessingResultBuilder` that is used by the state machines.

Happy to discuss this further next week.

**Comment by Zelldon** [comment]
Let's discuss this further @pihme :)

 For me the builder wouldn't be part of the RecordProcessor interface only the return of the Result. How the result is created was in my opinion implementation detail, which can be done differently in engine, tests etc. But yes let's discuss this further :)

**Comment by pihme** [comment]
Well, that changes it of course. As long as the interface is not based on byte buffers, I don't have much of an objection.

Still, I would expect this interface to be implemented by the platform. Don't see why engine wants to implement it one way, and backup processor wants to implement it a different way. But maybe you can ask Deepthi what her intuition is on this. 

I think it's great that she is implementing another record processor. Because she can start from a green field and doesn't need to carry over preconceptions from the Engine side. Plus we have a validation that the abstraction works for more than just the Engine.

**Comment by Zelldon** [comment]
https://github.com/camunda/zeebe/pull/10188 removes the LogStreamWriter part

**Comment by Zelldon** [comment]
https://github.com/camunda/zeebe/pull/10196 removes the response part

**Comment by Zelldon** [comment]
This is now done :partying_face:
